File: bestia-zoneserver/src/main/java/net/bestia/zoneserver/actor/chat/ChatActor.java
package net.bestia.zoneserver.actor.chat;

import akka.actor.ActorRef;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import net.bestia.messages.chat.ChatMessage;
import net.bestia.zoneserver.actor.SpringExtension;
import net.bestia.zoneserver.actor.zone.ClientMessageDigestActor;
import net.bestia.zoneserver.chat.ChatCommandService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import java.util.Objects;

/**
 * This actor processes chat messages from the clients to the bestia system. It
 * will check the preconditions and redirect it to the appropriate receivers
 * which will then in turn work with the chat message or redirect it to the chat
 * command system.
 * 
 * @author Thomas Felix
 *
 */
@Component
@Scope("prototype")
public class ChatActor extends ClientMessageDigestActor {

	private final LoggingAdapter LOG = Logging.getLogger(getContext().system(), this);
	public static final String NAME = "chat";

	private final ChatCommandService chatCmdService;

	private ActorRef publicChatActor;
	private ActorRef whisperChatActor;
	private ActorRef guildChatActor;
	private ActorRef partyChatActor;

	@Autowired
	public ChatActor(ChatCommandService chatCmdService) {

		this.chatCmdService = Objects.requireNonNull(chatCmdService);

		publicChatActor = SpringExtension.actorOf(getContext(), PublicChatActor.class);
		whisperChatActor = SpringExtension.actorOf(getContext(), WhisperChatActor.class);
		partyChatActor = SpringExtension.actorOf(getContext(), PartyChatActor.class);
		guildChatActor = SpringExtension.actorOf(getContext(), GuildChatActor.class);
		this.redirectConfig.match(ChatMessage.class, this::onChatMessage);
	}

	private void onChatMessage(ChatMessage chatMsg) {
		if (chatCmdService.isChatCommand(chatMsg.getText())) {
			chatCmdService.executeChatCommand(chatMsg.getAccountId(), chatMsg.getText());
			return;
		}

		switch (chatMsg.getChatMode()) {
		case PUBLIC:
			publicChatActor.tell(chatMsg, getSelf());
			break;
		case WHISPER:
			whisperChatActor.tell(chatMsg, getSelf());
			break;
		case PARTY:
			partyChatActor.tell(chatMsg, getSelf());
			break;
		case GUILD:
			guildChatActor.tell(chatMsg, getSelf());
			break;

		default:
			LOG.warning("Message type not yet supported.");
			break;
		}
	}
}

File: bestia-zoneserver/src/main/kotlin/net/bestia/zoneserver/actor/chat/ChatActor.kt
package net.bestia.zoneserver.actor.chat

import mu.KotlinLogging
import net.bestia.messages.chat.ChatMode
import net.bestia.messages.chat.ChatRequest
import net.bestia.zoneserver.actor.Actor
import net.bestia.zoneserver.actor.SpringExtension
import net.bestia.zoneserver.actor.routing.DynamicMessageRoutingActor
import net.bestia.zoneserver.chat.ChatCommandService

private val LOG = KotlinLogging.logger { }

/**
 * This actor processes chat messages from the clients to the bestia system. It
 * will check the preconditions and redirect it to the appropriate receivers
 * which will then in turn work with the chat message or redirect it to the chat
 * command system.
 *
 * @author Thomas Felix
 */
@Actor
class ChatActor(
    private val chatCmdService: ChatCommandService
) : DynamicMessageRoutingActor() {

  private val publicChatActor = SpringExtension.actorOf(context, PublicChatActor::class.java)
  private val whisperChatActor = SpringExtension.actorOf(context, WhisperChatActor::class.java)
  private val guildChatActor = SpringExtension.actorOf(context, GuildChatActor::class.java)
  private val partyChatActor = SpringExtension.actorOf(context, PartyChatActor::class.java)

  override fun createReceive(builder: BuilderFacade) {
    builder.matchRedirect(ChatRequest::class.java, this::onChatMessage)
  }

  private fun onChatMessage(chat: ChatRequest) {
    LOG.trace { "Received: $chat" }

    if (chatCmdService.isChatCommand(chat.text)) {
      chatCmdService.executeChatCommand(chat.accountId, chat.text)
      return
    }

    when (chat.chatMode) {
      ChatMode.PUBLIC -> publicChatActor.tell(chat, self)
      ChatMode.WHISPER -> whisperChatActor.tell(chat, self)
      ChatMode.PARTY -> partyChatActor.tell(chat, self)
      ChatMode.GUILD -> guildChatActor.tell(chat, self)
      else -> {
        LOG.warn { "Message type ${chat.chatMode} not yet supported." }
        unhandled(chat)
      }
    }
  }

  companion object {
    const val NAME = "chat"
  }
}